File: src/leetcodejava/array/TwoSum1.java
package leetcodejava.array;

import org.junit.Assert;
import org.junit.Test;

import java.util.HashMap;
import java.util.Map;

/**
 * This is the solution of No. 1 problem in the LeetCode,
 * the website of the problem is as follow:
 * https://leetcode-cn.com/problems/two-sum
 * <p>
 * The description of problem is as follow:
 * ==========================================================================================================
 * 给定一个整数数组 nums 和一个目标值 target，请你在该数组中找出和为目标值的那 两个 整数，并返回他们的数组下标。
 * <p>
 * 你可以假设每种输入只会对应一个答案。但是，数组中同一个元素不能使用两遍。
 * <p>
 * 示例:
 * <p>
 * 给定 nums = [2, 7, 11, 15], target = 9
 * <p>
 * 因为 nums[0] + nums[1] = 2 + 7 = 9
 * 所以返回 [0, 1]
 * <p>
 * 来源：力扣（LeetCode）
 * ==========================================================================================================
 *
 * @author zhangyu (zhangyuyu417@gmail.com)
 */
public class TwoSum1 {
    @Test
    public void twoSumTest() {
        int[] nums = {2, 7, 11, 15};
        int target = 9;
        int[] result = twoSum(nums, target);
        Assert.assertEquals(result.length, 2);
    }

    /**
     * 一种比较优的解决方式 利用hashmap,时间复杂度:O(n) 空间复杂度:(1)
     *
     * @param nums   数组
     * @param target 目标
     * @return 数组
     */
    private int[] twoSum(int[] nums, int target) {
        if (nums == null || nums.length < 2) {
            return new int[]{-1, -1};
        }
        Map<Integer, Integer> map = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            if (map.containsKey(target - nums[i])) {
                return new int[]{map.get(target - nums[i]), i};
            } else {
                map.put(nums[i], i);
            }
        }
        return new int[]{-1, -1};
    }
}

File: src/leetcodepython/array/two_sum_1.py
# -*- coding:utf-8 -*-
'''
/**
 * This is the solution of No. 1 problem in the LeetCode,
 * the website of the problem is as follow:
 * https://leetcode-cn.com/problems/two-sum
 *
 * The description of problem is as follow:
 * ==========================================================================================================
 * 给定一个整数数组 nums 和一个目标值 target，请你在该数组中找出和为目标值的那 两个 整数，并返回他们的数组下标。
 *
 * 你可以假设每种输入只会对应一个答案。但是，数组中同一个元素不能使用两遍。
 *
 * 示例:
 *
 * 给定 nums = [2, 7, 11, 15], target = 9
 *
 * 因为 nums[0] + nums[1] = 2 + 7 = 9
 * 所以返回 [0, 1]
 *
 * 来源：力扣（LeetCode）
 * ==========================================================================================================
 *
 * @author zhangyu (zhangyuyu417@gmail.com)
 */
'''
from typing import List


class Solution:
    def two_sum(self, nums: List[int], target: int) -> int:
        '''
            计算组成和下标
        Args:
            nums: 数组
            target: 目标值
        Returns:
            目标值下标
        '''
        hashtable = {}
        for i in range(len(nums)):
            if (target - nums[i]) in hashtable:
                return [hashtable[target - nums[i]], i]
            else:
                hashtable[nums[i]] = i
        return [-1, -1]


if __name__ == '__main__':
    nums = [2, 5, 7, 11]
    target = 7
    solution = Solution()
    result = solution.two_sum(nums, target)
    print(result)
    assert result == [0, 1]
